third_party/terraform/resources/resource_google_project_services.go

```diff
@@ -48,330 +54,321 @@ func resourceGoogleProjectServices() *schema.Resource {
-			}
-			if err != nil {
-				// Check for a "precondition failed" error. The API seems to randomly
-				// (although more than 50%) return this error when enabling certain
-				// APIs. It's transient, so we catch it and re-raise it as an error that
-				// is retryable instead.
-				if gerr, ok := err.(*googleapi.Error); ok {
-					if (gerr.Code == 400 || gerr.Code == 412) && gerr.Message == "Precondition check failed." {
-						return &googleapi.Error{
-							Code:    503,
-							Message: "api returned \"precondition failed\" while enabling service",
-						}
-					}
-				}
-				return errwrap.Wrapf("failed to issue request: {{err}}", err)
-			}
-
-			// Poll for the API to return
-			activity := fmt.Sprintf("apis %q to be enabled for %s", services, pid)
-			waitErr := serviceUsageOperationWait(config, sop, activity)
-			if waitErr != nil {
-				return waitErr
-			}
-
-			// Accumulate the list of services that are enabled on the project
-			enabledServices, err := getApiServices(pid, config, nil)
-			if err != nil {
-				return err
-			}
-
-			// Diff the list of requested services to enable against the list of
-			// services on the project.
-			missing := diffStringSlice(services, enabledServices)
-
-			// If there are any missing, force a retry
-			if len(missing) > 0 {
-				// Spoof a googleapi Error so retryTime will try again
-				return &googleapi.Error{
-					Code:    503,
-					Message: fmt.Sprintf("The service(s) %q are still being enabled for project %s. This isn't a real API error, this is just eventual consistency.", missing, pid),
-				}
-			}
-
+		nextBatch := services[i:j]
+		if len(nextBatch) == 0 {
+			// All batches finished, return.
 			return nil
-		}, 10); err != nil {
-			return errwrap.Wrap(err, fmt.Errorf("failed to enable service(s) %q for project %s", services, pid))
 		}
+
+		if err := sendServiceUsageEnableServicesRequest(nextBatch, project, config); err != nil {
+			return err
+		}
+		log.Printf("[DEBUG] Finished enabling next batch of %d project services: %+v", len(nextBatch), nextBatch)
 	}
 
-	return nil
+	log.Printf("[DEBUG] Verifying that all services are enabled")
+	return waitForServiceUsageEnabledServices(services, project, config)
 }
 
-func diffStringSlice(wanted, actual []string) []string {
-	var missing []string
-
-	for _, want := range wanted {
-		found := false
+// waitForServiceUsageEnabledServices doesn't resend enable requests - it just
+// waits for service enablement status to propagate. Essentially, it waits until
+// all services show up as enabled when listing services on the project.
+func waitForServiceUsageEnabledServices(services []string, project string, config *Config) error {
+	missing := make([]string, 0, len(services))
+	delay := time.Duration(0)
+	interval := time.Second
+	err := retryTime(func() error {
+		// Get the list of services that are enabled on the project
+		enabledServices, err := getEnabledServiceSet(project, config, nil)
```

Comment:
Why don't we want to ignore the ignored project services here? Users can't set them due to https://github.com/GoogleCloudPlatform/magic-modules/blob/2177135b1395395416e760d9ede71395f9f1777d/third_party/terraform/resources/resource_google_project_services.go#L45

I don't see how they'd ever end up in `services`?

Reply:
we're actually ignoring them when getting the enabled services - however, i don't really know why we have it as an argument, considering we don't ever call it with a different value. I'm going to remove the argument.

Reply:
We call it with the same value everywhere but here, where it's `nil`. So on this call, it's not being ignored, right?

Reply:
that was a typo by me